client/pluginmanager/drivermanager/manager.go

```diff
@@ -0,0 +1,334 @@
+	// drivers
+	AllowedDrivers map[string]struct{}
+
+	// BlockedDrivers if set will not allow the given driver plugins to start
+	BlockedDrivers map[string]struct{}
+}
+
+// manager is used to manage a set of driver plugins
+type manager struct {
+	// logger is the logger used by the device manager
+	logger log.Logger
+
+	// state is used to manage the device managers state
+	state StateStorage
+
+	// ctx is used to shutdown the device manager
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// loader is the plugin loader
+	loader loader.PluginCatalog
+
+	// pluginConfig is the config passed to the launched plugins
+	pluginConfig *base.ClientAgentConfig
+
+	// updater is used to update the node when device information changes
+	updater UpdateNodeDriverInfoFn
+
+	// instances is the list of managed devices, access is serialized by instanceMu
+	instances   map[string]*instanceManager
+	instancesMu sync.RWMutex
+
+	// reattachConfigs stores the plugin reattach configs
+	reattachConfigs    map[loader.PluginID]*shared.ReattachConfig
+	reattachConfigLock sync.Mutex
+
+	// allows/block lists
+	allowedDrivers map[string]struct{}
+	blockedDrivers map[string]struct{}
+
+	// readyCh is ticked once at the end of Run()
+	readyCh chan struct{}
+}
+
+// New returns a new driver manager
+func New(c *Config) *manager {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &manager{
+		logger:          c.Logger.Named("driver_mgr"),
+		state:           c.State,
+		ctx:             ctx,
+		cancel:          cancel,
+		loader:          c.Loader,
+		pluginConfig:    c.PluginConfig,
+		updater:         c.Updater,
+		instances:       make(map[string]*instanceManager),
+		reattachConfigs: make(map[loader.PluginID]*shared.ReattachConfig),
+		allowedDrivers:  c.AllowedDrivers,
+		blockedDrivers:  c.BlockedDrivers,
+		readyCh:         make(chan struct{}),
+	}
+}
+
+// PluginType returns the type of plugin this mananger mananges
+func (*manager) PluginType() string { return base.PluginTypeDriver }
+
+// Run starts the mananger, initializes driver plugins and blocks until Shutdown
+// is called.
+func (m *manager) Run() {
+	// Load any previous plugin reattach configuration
+	m.loadReattachConfigs()
+
+	// Get driver plugins
+	driversPlugins := m.loader.Catalog()[base.PluginTypeDriver]
+	if len(driversPlugins) == 0 {
+		m.logger.Debug("exiting since there are no driver plugins")
+		m.cancel()
+		return
+	}
+
+	var skippedDrivers []string
+	for _, d := range driversPlugins {
+		id := loader.PluginInfoID(d)
+		// Skip drivers that are not in the allowed list if it is set.
+		if _, ok := m.allowedDrivers[id.Name]; len(m.allowedDrivers) > 0 && !ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+		// Skip fingerprinting drivers that are in the blocked list
+		if _, ok := m.blockedDrivers[id.Name]; ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
```

Comment:
What happens if you have a driver in the reattach configs but that has subsequently been blacklisted?